src/common/config_proxy.h

```diff
@@ -0,0 +1,153 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+
+#pragma once
+
+#include <memory>
+#include <type_traits>
+
+#include "common/config.h"
+#include "common/config_fwd.h"
+
+class ConfigProxy {
+  /**
+   * The current values of all settings described by the schema
+   */
+
+  ConfigValues values;
+  std::unique_ptr<md_config_t> config;
```

Comment:
why is heap allocation necessary for the md_config_t part?

Reply:
@cbodley right, it's not needed. i will make `ConfigProxy` a template, and specialize it with `LockPolicy` so that the plain `ConfigProxy<LockPolicy::MUTEX>` will use a `config` on stack, while `ConfigProxy` will be a `sharded<>` service, and the ConfigProxy shards will share the same `md_config_t` instance by holding a read-only `const md_config_impl<LockPolicy::SINGLE>*`, only the shard.0 will have a local copy of `unique_ptr<md_config_impl<LockPolicy::SINGLE>>`.

i can address your comment in this PR or in a follow-up PR. but i am inclined to have it in the follow-up PR. as it's already included by Sage's test batch. what do you think?

Reply:
okay. the cost of allocating here isn't going to be an issue unless we're constantly creating/destroying these, but it sounds like this constructor will only run at startup